Support creating and updating users with a PIN

jm-araujo opened the pull request:
Currently the provider does not support creating or updating users with a PIN number. Users are always created without a pin and if a pin is desired it must be set in some other way.

The provider declares a `User` structure in `user.go` which is used in all user operations. Because this struct has a PIN field as a `bool`, it can correctly map the Pritunl API responses to GET requests (which merely describe whether the user has or not a PIN), but it cannot be used to POST or PUT users with an actual PIN.

This PR adds a new structure `NewUser` which is a copy of `User` but with the PIN field as a `string`. Additionally, the relevant interfaces in the client for creating and updating users are also updated to handle and send this `NewUser`.

disc reviewed (changes requested):
Hello @jm-araujo. Let's try to avoid creating a copy of `User` struct.
Probably you can add a new attribute for `User` struct that has a similar name and sets/updates pin only when is not empty, but `pin` property should contain a bool value and reflect the state that Pin has already been set.

Also, please add an example of setting a new attribute into an `example` directory and don't forget to write some tests that covers your changes.

jm-araujo commented:
Hi @disc 

Adding a new attribute to User while keeping the old one didn't seem viable, or at least I couldn't find a way to do it without implementing custom marshaling functions for the whole `User` struct. So instead I changed the `User.Pin` field type to use a new struct `Pin` and added custom marshaling functions just for that struct (which consider the string value when encoding and the boolean value when decoding).

As for tests, I've updated the existing User acceptance test to consider the pin attribute as well.

wdyt?

disc commented on internal/pritunl/user.go:
```diff
@@ -26,10 +29,43 @@ type User struct {
+// When marshaling a User JSON, the "pin" field will contain the PIN secret
+// if it is set, otherwise the field is excluded. This is used when making
+// a user create or update request to the Pritunl API.
+func (p Pin) MarshalJSON() ([]byte, error) {
```
Add a pointer for a struct name in a receiver

jm-araujo replied:
Changing the receiver from `Pin` to `*Pin` here breaks the existing User tests. `Pin` is a small struct so as I understand it there shouldn't be an advantage to using a pointer no?

jm-araujo replied:
Actually took a better look at this and got it working alright. ✅

disc commented on internal/pritunl/user.go:
```diff
@@ -26,10 +29,43 @@ type User struct {
 }
+
+type Pin struct {
+	IsSet  bool
```
I don't see any read of the field `IsSet`, only two write calls. Do you really need it?

jm-araujo replied:
I don't think so. This was added just to keep the functionality previously provided by the `Pin bool` field, which I understood was to be kept.

disc commented on examples/provider/provider.tf:
```diff
@@ -28,6 +28,7 @@ resource "pritunl_user" "test" {
   name            = "test-user"
   organization_id = pritunl_organization.developers.id
   email           = "test@test.com"
+  pin             = "123456"
```
Create an additional example instead of modifying this one. Pin usage is an optional (but secure) way for users management

jm-araujo replied:
Done.

disc commented on internal/provider/resource_user.go:
```diff
@@ -170,6 +177,13 @@ func resourceUserUpdate(ctx context.Context, d *schema.ResourceData, meta interf
 
+	if d.HasChange("pin") {
+		if v, ok := d.GetOk("pin"); ok {
+			user.Pin.IsSet = false
```
Looks like no one read `IsSet` property, why do you reset it? Probably I missed something?

jm-araujo replied:
This was a relic from an earlier iteration. Removed it.

disc commented on internal/provider/resource_user_test.go:
```diff
@@ -32,15 +34,16 @@ func TestAccPritunlUser(t *testing.T) {
 }
 
-func testPritunlUserConfig(username, orgName string) string {
+func testPritunlUserConfig(username, orgName, pin string) string {
```
Create an additional config helper and a new test case instead of modifying this one, because pin is an optional value for a user, and it's better to have a new test case for this scenario.

jm-araujo replied:
Added a new test case and helper function.

disc reviewed (changes requested):
Looks indeed better than the previous solution. Let's fix minor issues and proceed to the release of your contribution.

disc commented on internal/provider/resource_user_test.go:
```diff
@@ -30,17 +31,52 @@ func TestAccPritunlUser(t *testing.T) {
+    organization_id = pritunl_organization.test.id
+    `, username, orgName)
+
+	if pin != "" {
```
Let's remove this if condition and assume that the PIN is always set. it should make a template more easy and clear

disc commented:
@jm-araujo 
Thanks for your contribution. I'm going to merge and release these changes soon.

jm-araujo commented:
Glad to hear it. Thanks!
